radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/addressbook/AddressBook.java

```diff
@@ -91,35 +93,95 @@
  * Manages known peers network addresses and their metadata.
  */
 public final class AddressBook {
-	private static final Comparator<AddressBookEntry.PeerAddressEntry> entryComparator = (a, b) -> {
-		final var aLastSuccess = a.getLastSuccessfulConnection().orElse(Instant.MIN);
-		final var bLastSuccess = b.getLastSuccessfulConnection().orElse(Instant.MIN);
-		return -aLastSuccess.compareTo(bLastSuccess);
-	};
 
-	private final NodeId selfNodeId;
+	/**
+	 * A stateful comparator for known peer addresses that uses both their latest connection status (persistent)
+	 * and a number of failed connection attempts (volatile). Failure counts are used to cycle the addresses for retries.
+	 * The entries are sorted in the following manner:
+	 * successful -> unknown (no connection attempted yet) -> failed (by num of failures)
+	 */
+	private static final class PeerAddressEntryComparator implements Comparator<PeerAddressEntry> {
+		private final Map<RadixNodeUri, Integer> failureCounts = new HashMap<>();
```

Comment:
What do you think about moving this to `PeerAddressEntry`? Not sure it is the Comparator responsibility.

Reply:
@LukasGasior1 answer: I was considering adding it to the address book entry and persisting (failureCount), but I think it's not really required to be persisted and keeping it in memory saves a db call